### PR TITLE
add catch for TimeoutException

### PR DESCRIPTION
**Pull request by rgildein**
Add catch for TimeoutException and enable timeout configuration via environmnet variable. Use retry decorator in non-invasive functions.

related: #103

### CODE REVIEW

**Review by agileshaw — COMMENTED** [review]
Maybe I missed some discussions on this, but I don't quite understand the reason of having `CommandExecutionFailed` and `ActionFailedToRun`. What is the behavior exactly when running actions or commands in `retry` hook?

**Comment by rgildein** [comment]
> Maybe I missed some discussions on this, but I don't quite understand the reason of having `CommandExecutionFailed` and `ActionFailedToRun`. What is the behavior exactly when running actions or commands in `retry` hook?

It's only to be able to catch it and not retry any action or command, only retry if waiting for action failed or getting action results failed, etc. We do not want to trigger single action multiple times, it can be dangerous.

**Review by agileshaw — APPROVED** [review]
LGTM

File: cou/utils/juju_utils.py [hunk 1]
```diff
@@ -270,24 +298,20 @@ async def run_action(
         :param action_params: Dictionary of config options for action, defaults to None
         :type action_params: Optional[Dict], optional
         :param raise_on_failure: Raise ActionFailed exception on failure, defaults to False
-        :type raise_on_failure: bool, optional
+        :type raise_on_failure: bool
+        :raises UnitNotFound: _description_
```
Comment (gabrielcocenza) [inline cou/utils/juju_utils.py:302]: Missing a simple description
```suggestion
        :raises UnitNotFound: When cannot find a valid unit.
```

File: cou/utils/juju_utils.py [hunk 2]
```diff
@@ -254,6 +258,30 @@ async def get_status(self) -> FullStatus:
...
+    async def _get_action_result(self, action: Action, raise_on_failure: bool) -> Action:
+        """Get results from action.
+
+        :param action: Action object
+        :type: Action
+        :param raise_on_failure: Raise ActionFailed exception on failure, defaults to False
+        :type raise_on_failure: bool
+        :return: action results
+        :rtype: Action
+        :raises ActionFailed: _description_
```
Comment (gabrielcocenza) [inline cou/utils/juju_utils.py:271]:
```suggestion
        :raises ActionFailed: When cannot run an application action.
```

File: cou/utils/juju_utils.py [hunk 3]
```diff
@@ -36,11 +38,11 @@
...
 
 JUJU_MAX_FRAME_SIZE: int = 2**30
-DEFAULT_TIMEOUT: int = 60
+DEFAULT_TIMEOUT: int = int(os.environ.get("COU_TIMEOUT", 10))
 DEFAULT_MAX_WAIT: int = 5
 DEFAULT_WAIT: float = 1.1
-DEFAULT_MODEL_RETRIES: int = 30
-DEFAULT_MODEL_RETRY_BACKOFF: int = 2
+DEFAULT_MODEL_RETRIES: int = int(os.environ.get("COU_MODEL_RETRIES", 5))
+DEFAULT_MODEL_RETRY_BACKOFF: int = int(os.environ.get("COU_MODEL_RETRY_BACKOFF", 2))
```
Comment (gabrielcocenza) [inline cou/utils/juju_utils.py:45]: Can we open an issue to document those env vars just to not lose track?

File: cou/utils/juju_utils.py [hunk 4]
```diff
@@ -301,14 +325,15 @@ async def run_on_unit(
         :type timeout: Optional[int]
         :returns: action.data['results'] {'Code': '', 'Stderr': '', 'Stdout': ''}
         :rtype: Dict[str, str]
-        :raises: UnitNotFound
-        :raises: ActionFailed
+        :raises UnitNotFound: _description_
+        :raises ActionFailed: _description_
```
Comment (gabrielcocenza) [inline cou/utils/juju_utils.py:329]: provide a simple description

File: tests/unit/utils/test_juju_utils.py [hunk 5]
```diff
@@ -304,40 +304,49 @@ async def test_coumodel_get_status(mocked_model):
 
 
 @pytest.mark.asyncio
-async def test_coumodel_run_action(mocked_model):
+async def test_coumodel_get_action_result(mocked_model):
     """Test COUModel run action."""
```
Comment (gabrielcocenza) [inline tests/unit/utils/test_juju_utils.py:308]:
```suggestion
    """Test COUModel get action result."""
```

File: tests/unit/utils/test_juju_utils.py [hunk 6]
```diff
@@ -304,40 +304,49 @@ async def test_coumodel_get_status(mocked_model):
...
     mocked_action.wait.assert_awaited_once_with()
-    assert action == mocked_result
+    assert action == mocked_action.wait.return_value
+    mocked_model.get_action_status.assert_not_awaited()
 
 
 @pytest.mark.asyncio
-async def test_coumodel_run_action_failure(mocked_model):
+async def test_coumodel_get_action_result_failure(mocked_model):
     """Test COUModel run action failing."""
```
Comment (gabrielcocenza) [inline tests/unit/utils/test_juju_utils.py:321]:
```suggestion
    """Test COUModel get action result failing."""
```

**Review by gabrielcocenza — COMMENTED** [review]
Small details, otherwise LGTM

File: cou/utils/juju_utils.py (hunk 1 above)
Comment (rgildein) [reply cou/utils/juju_utils.py:302]: I saw usage of `_description_` in multiple places, do you know what is actually doing? I was thinking that it's using docstring from exception, but I could not found any information about it.

File: cou/utils/juju_utils.py (hunk 3 above)
Comment (rgildein) [reply cou/utils/juju_utils.py:45]: I working on documentation so I will mention it there. Not sure if these two are needed, but I can add them.

Comment (gabrielcocenza) [reply cou/utils/juju_utils.py:45]: As discussed yesterday, you said that you are in charge for documentation and you won't miss this, so resolve the conversation

File: cou/utils/juju_utils.py (hunk 1 above)
Comment (gabrielcocenza) [reply cou/utils/juju_utils.py]: As discussed yesterday, this is automatically generated from a plugin that I use to generate docstrings. We should replace all `_description_` later.

File: cou/utils/juju_utils.py (hunk 4 above)
Comment (rgildein) [reply cou/utils/juju_utils.py]: You are right, I tried to be fast and did not realized that it's `ApplicationFailed` and not `ApplicationNotFound`.

Comment (rgildein) [reply cou/utils/juju_utils.py]: What do you say about this `When the application status is in error (it's not 'completed').`?

Comment (gabrielcocenza) [reply cou/utils/juju_utils.py]: Sure, let's try to be consistent and change to this description on all `ApplicationFailed`

**Review by gabrielcocenza — DISMISSED** [review]

File: cou/utils/juju_utils.py [hunk 7]
```diff
@@ -301,15 +325,16 @@ async def run_on_unit(
         :type timeout: Optional[int]
         :returns: action.data['results'] {'Code': '', 'Stderr': '', 'Stdout': ''}
         :rtype: Dict[str, str]
-        :raises: UnitNotFound
-        :raises: ActionFailed
+        :raises UnitNotFound: When cannot find a valid unit.
```
Comment (aieri) [inline cou/utils/juju_utils.py:328]: Nit:
```suggestion
        :raises UnitNotFound: When a valid unit cannot be found. 
```

File: tests/unit/test_cli.py [hunk 8]
```diff
@@ -187,6 +187,7 @@ async def test_entrypoint_commands(mocker, command, function_call):
         (Exception("An error occurred"), "2"),
         (COUException("Caught error"), "1"),
         (UnitNotFound("Unit not found"), "1"),
+        (TimeoutException("Connection timeouted"), "1"),
```
Comment (aieri) [inline tests/unit/test_cli.py:190]: Nit:
```suggestion
        (TimeoutException("The connection timed out"), "1"),
```

**Review by aieri — APPROVED** [review]
LGTM, w/some small rewording suggestions

**Review by aieri — APPROVED** [review]

**Review by gabrielcocenza — APPROVED** [review]

**Pull request merged**